adv-rumors.lua

```diff
@@ -1,83 +1,139 @@
+    default_pos={x=25,y=25},
+    default_enabled=true,
+    viewscreens='dungeonmode/Conversation',
+    overlay_onupdate_max_freq_seconds=0,
+}
+
+function AdvRumorsOverlay:overlay_onupdate()
+    rumorUpdate()
+end
+
+OVERLAY_WIDGETS = {message=AdvRumorsOverlay}
+
+-- CORE FUNCTIONS
+
+-- Converts the choice's print string to a lua string
+function choiceToString(choice)
+    local line_table = {}
+    for i, data in ipairs(choice.print_string.text) do
+        table.insert(line_table, data.value)
+    end
+    return table.concat(line_table, "\n")
+end
+
+-- Renames the choice's print string based on string, respecting the newlines
+function renameChoice(text, choice)
+    -- Clear the string
+    for i, data in ipairs(choice.print_string.text) do
+        df.delete(data)
+    end
+    choice.print_string.text:resize(0)
+
+    -- Split the string assuming \n is newline
+    local line_table = string.split(text, "\n")
+    for i, line in ipairs(line_table) do
+        -- Create a df string for each line
+        local line_ptr = df.new('string')
+        line_ptr.value = line
+        -- Insert it into the text data
+        choice.print_string.text:insert('#', line_ptr)
     end
 end
 
 function addKeyword(choice, keyword)
+    -- Prevent duplicate keywords
+    for i, kword in ipairs(choice.key_word) do
+        if kword.value == keyword then
+            return
+        end
+    end
     local keyword_ptr = df.new('string')
     keyword_ptr.value = keyword
-    choice.keywords:insert('#', keyword_ptr)
+    choice.key_word:insert('#', keyword_ptr)
 end
 
-function rumorUpdate()
-    local improveReadability = true
-    local addKeywordSlew = true
-    local shortenString = true
-    local addKeywordNames = true
-
-    for i, choice in ipairs(df.global.adventure.conversation.choices) do
-        if choice.choice.type == df.talk_choice_type.SummarizeConflict then
-            if improveReadability then
-                condenseChoiceTitle(choice)
-            end
-            if shortenString then
-                condenseChoiceTitle(choice)
-                choice.title[0].value = choice.title[0].value
-                    :gsub("Summarize the conflict in which +", "")
-                    :gsub("This occurred +", "")
-            end
-            if addKeywordSlew then
-                if string.find(choice.title[0].value, "slew") then
-                    addKeyword(choice, 'slew')
-                end
-            end
-            if addKeywordNames then
-                local title = choice.title[0].value
-                for keyword in title:sub(1, title:find('%(') - 1):gmatch('%w+') do
-                    keyword = dfhack.utf2df(dfhack.df2utf(keyword):lower())
-                    if not names_blacklist[keyword] then
-                        addKeyword(choice, keyword)
-                    end
-                end
-            end
+function addKeywordsForChoice(choice)
+    local fulltext = choiceToString(choice)
```

Comment:
It's far easier to process the title as a string in the function, simplifies code tbh

Reply:
I'm not sure that's true. it would require the current code to be *refactored*, certainly, but I don't think it would be any more complex. String copying is expensive, and though we're fine with running this every frame, there's no need to be wasteful.